### PR TITLE
feat: steering to avoid chain breaks and clashes

### PR DESCRIPTION
**Pull request by ludwigwinkler**
Add Steering functionality to BioEmu

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
 
 This code only supports sampling structures of monomers. You can try to sample multimers using the [linker trick](https://x.com/ag_smith/status/1417063635000598528), but in our limited experiments, this has not worked well.
 
+## Steering structures
+
+BioEmu includes a steering system that uses Sequential Monte Carlo (SMC) to guide the diffusion process toward more physically plausible protein structures. Steering applies potential energy functions during denoising to favor conformations that satisfy physical constraints. Algorithmically, steering simulates *multiple particles* per desired sample and resamples between these particles according to the favorability of the provided potentials. 
+
+### Quick Start with Steering
```
Comment (sarahnlewis) [inline README.md]:
```suggestion
### Quick start with steering
```

File: README.md [hunk 2]
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
...
+
+### Key Steering Parameters
+
+- `num_steering_particles`: Number of particles per sample (1 = no steering, >1=steering)
+- `steering_start_time`: When to start steering (0.0-1.0, default: 0.0)
+- `steering_end_time`: When to stop steering (0.0-1.0, default: 1.0)
+- `resampling_freq`: How often to resample particles (default: 1)
+- `steering_potentials_config`: Path to potentials configuration file (optional, defaults to physical_potentials.yaml)
+
+### Available Potentials
```
Comment (sarahnlewis) [inline README.md]:
```suggestion
### Available potentials
```

File: README.md [hunk 3]
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
 
 This code only supports sampling structures of monomers. You can try to sample multimers using the [linker trick](https://x.com/ag_smith/status/1417063635000598528), but in our limited experiments, this has not worked well.
 
+## Steering structures
+
+BioEmu includes a steering system that uses Sequential Monte Carlo (SMC) to guide the diffusion process toward more physically plausible protein structures. Steering applies potential energy functions during denoising to favor conformations that satisfy physical constraints. Algorithmically, steering simulates *multiple particles* per desired sample and resamples between these particles according to the favorability of the provided potentials.
```
Comment (sarahnlewis) [inline README.md]: I suggest including a reference. Also, spell out that no guidance term is added to the state updates; we only do resampling.

**Review by sarahnlewis — COMMENTED** [review]
I started leaving some nitty comments but then realised review is maybe premature. Let's discuss 1:1.

Maybe it would make sense to split this into two PRs: 1. add hydra interface, 2. add steering.

I'd suggest returning all samples from each group of particles instead of following [boltz]( https://github.com/jwohlwend/boltz/blob/cb04aeccdd480fd4db707f0bbafde538397fa2ac/src/boltz/model/modules/diffusionv2.py#L486) and only returning 1 sample from each. It will be less wasteful and will make the logic around 'batch size' less confusing.  Also, because you use stratified resampling, there might be more useful diversity within each particle group than there is there.

File: src/bioemu/config/denoiser/dpm.yaml [hunk 4]
```diff
@@ -2,5 +2,5 @@ _target_: bioemu.shortcuts.dpm_solver
 _partial_: true
 eps_t: 0.001
 max_t: 0.99
-N: 50
-noise: 0.0
+N: 100
+noise: 0.5 # original dpm =0 for ode
```
Comment (YuuuXie) [inline src/bioemu/config/denoiser/dpm.yaml]: Do you find when the noise is increased to 0.5 it does not work with 50 steps?

Comment (YuuuXie) [reply src/bioemu/config/denoiser/dpm.yaml]: I'd suggest to keep the original setting (N=50, noise=0) as default. On the other hand, we can override the `denoiser` setting in `bioemu.yaml` if we use it as a steering example

File: pyproject.toml [hunk 5]
```diff
@@ -22,6 +22,7 @@ dependencies = [
     "stackprinter",
     "typer",
     "uv",
+    "einops"
```
Comment (YuuuXie) [inline pyproject.toml]: I don't see `einops` is used anywhere, do we need it?

File: src/bioemu/convert_chemgraph.py [hunk 6]
```diff
@@ -2,12 +2,16 @@
 # Licensed under the MIT License.
 import logging
 from pathlib import Path
+import os
 
+from matplotlib.pylab import f
```
Comment (YuuuXie) [inline src/bioemu/convert_chemgraph.py]: remove unused imports

File: src/bioemu/convert_chemgraph.py [hunk 7]
```diff
@@ -462,11 +702,17 @@ def save_pdb_and_xtc(
 
     # .pdb files contain coordinates in Angstrom
     _write_pdb(
-        pos=pos_angstrom[0],
-        node_orientations=node_orientations[0],
+        pos=pos_angstrom[-1],
+        node_orientations=node_orientations[-1],
```
Comment (YuuuXie) [inline src/bioemu/convert_chemgraph.py]: why does it save the last frame instead of the first? (I guess in terms of topology it does not matter, but just curious if anything does not work)

File: src/bioemu/convert_chemgraph.py [hunk 8]
```diff
@@ -477,11 +723,23 @@ def save_pdb_and_xtc(
...
+            f"Filtered {num_samples_unfiltered} samples down to {len(traj)} "
+            "based on structure criteria. Filtering can be disabled with `--filter_samples=False`."
+        )
+        print(
+            f"Filtered {num_samples_unfiltered} samples down to {len(traj)} ",
+            "based on structure criteria. Filtering can be disabled with `--filter_samples=False`.",
+        )
+        # Filtering ratio computed but not logged
+
         filtered_traj = filter_unphysical_traj(traj)
```
Comment (YuuuXie) [inline src/bioemu/convert_chemgraph.py]: `filter_unphysical_traj` are called twice (line 732 and 743), and the logging/printing is duplicated

File: src/bioemu/denoiser.py [hunk 9]
```diff
@@ -1,16 +1,37 @@
...
 
-from .chemgraph import ChemGraph
-from .sde_lib import SDE, CosineVPSDE
-from .so3_sde import SO3SDE, apply_rotvec_to_rotmat
+import copy
+from torch_geometric.data.batch import Batch
+import time
+import torch.autograd.profiler as profiler
+from torch.profiler import profile, ProfilerActivity, record_function
+from tqdm.notebook import tqdm
```
Comment (YuuuXie) [inline src/bioemu/denoiser.py]: Had an impression this `tqdm.notebook` only works within the notebook, but maybe I'm wrong. It might be more general to just use `tqdm`

File: src/bioemu/denoiser.py [hunk 10]
```diff
@@ -67,7 +88,7 @@ def update_given_drift_and_diffusion(
         else:
             mean = x + drift * dt
             sample = mean + self.noise_weight * diffusion * torch.sqrt(dt.abs()) * z
-        return sample, mean
+        return sample, mean, z
```
Comment (YuuuXie) [inline src/bioemu/denoiser.py]: Do I understand correctly this `z` is not used anywhere, if so, we can just keep the original return without `z`?

File: src/bioemu/steering.py [hunk 11]
```diff
@@ -0,0 +1,873 @@
...
+import einops
+
+# No wandb logging needed
+
+import numpy as np
+import matplotlib.pyplot as plt
+
+
+@torch.enable_grad()
+def potential_gradient_minimization(x, potentials, learning_rate=0.1, num_steps=20):
```
Comment (YuuuXie) [inline src/bioemu/steering.py]: Shall we remove it if we don't have plan to include in this release?

**Review by YuuuXie — COMMENTED** [review]
Thanks for the great efforts! I left some comments, and still need to take a closer look at the steering code, potential definitions etc. 

Is there a plan of splitting the current one into two PRs as Sarah suggested? Or this is already one of them?

File: README.md [hunk 12]
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
...
+    resampling_freq=2
+)
+```
+
+### Key steering parameters
+
+- `num_steering_particles`: Number of particles per sample (1 = no steering, >1=steering)
+- `steering_start_time`: When to start steering (0.0-1.0, default: 0.0)
+- `steering_end_time`: When to stop steering (0.0-1.0, default: 1.0)
+- `resampling_freq`: How often to resample particles (default: 1)
```
Comment (sarahnlewis) [inline README.md]: Is this an interval or a frequency? Does resampling_freq=3 mean resampling after every 3 denoising steps?

File: README.md [hunk 13]
```diff
@@ -65,6 +66,54 @@ By default, unphysical structures (steric clashes or chain discontinuities) will
...
+- `steering_potentials_config`: Path to potentials configuration file (optional, defaults to physical_potentials.yaml)
+
+### Available potentials
+
+When steering is enabled (num_steering_particles > 1) and no additional `steering_potentials_config.yaml` is provided, BioEMU automatically loads `physical_potentials.yaml` by default, which includes:
```
Comment (sarahnlewis) [inline README.md]:
```suggestion
When steering is enabled (num_steering_particles > 1) and `steering_potentials_config` is not given, BioEmu automatically loads `physical_potentials.yaml` by default, which includes:
```
please `git grep BioEMU` and replace with `BioEmu`.

File: notebooks/disulfide_steering_example.py [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+    )
+    with open(config_path) as f:
+        potentials_config = yaml.safe_load(f)
+
+    # Enable/disable guidance steering on the disulfide potential
+    for key in potentials_config:
+        if key == "disulfide":
+            potentials_config[key]["guidance_steering"] = enable_guidance
+        else:
+            potentials_config[key]["guidance_steering"] = enable_guidance
```
Comment (sarahnlewis) [inline notebooks/disulfide_steering_example.py]: The 'if' and 'else' blocks are identical.

**Review by sarahnlewis — COMMENTED** [review]
Thanks for the new feature! I haven't finished reviewing yet but am leaving some comments on what I looked at so far.

File: src/bioemu/convert_chemgraph.py (hunk 7 above)
Comment (sarahnlewis) [reply src/bioemu/convert_chemgraph.py]: +1

File: .gitignore [hunk 15]
```diff
@@ -132,3 +134,22 @@ cython_debug/
...
+*/wandb/*
+*.npz
+*.pkl
+*amlt*
+*outputs*
+*cache*
+.cursor/
+.cursor/** */
+docs/*
+uv.lock
```
Comment (sarahnlewis) [inline .gitignore]: ??

File: .gitignore [hunk 16]
```diff
@@ -132,3 +134,22 @@ cython_debug/
...
+
+# samples
+*.pdb
```
Comment (sarahnlewis) [inline .gitignore]: This is problematic because there are .pdb files in the repo, referenced in tests. Instead of adding all this to .gitignore I suggest you don't save samples to the repo directory in the first place. Same comment for many of the other things added here.

File: README.md (hunk 12 above)
Comment (sarahnlewis) [reply README.md]: If resampling_freq=n means resampling every n denoising steps then this setting is _inverse_ of frequency, and is misnamed. You could call it resampling_interval instead.

File: src/bioemu/sample.py [hunk 17]
```diff
@@ -69,45 +75,119 @@ def main(
...
+        num_samples: Number of samples to generate. If `output_dir` already contains samples, this
+            function will only generate additional samples necessary to reach the specified `num_samples`.
+        output_dir: Directory to save the samples. Each batch of samples will initially be dumped
+            as .npz files. Once all batches are sampled, they will be converted to .xtc and .pdb.
+        batch_size_100: Batch size you'd use for a sequence of length 100. The batch size will be
+            calculated from this, assuming that the memory requirement to compute each sample scales
+            quadratically with the sequence length. A100-80GB would give you ~900 right at the memory
+            limit, so 500 is reasonable
+        model_name: Name of pretrained model to use. If this is set, you do not need to provide
+            `ckpt_path` or `model_config_path`. The model will be retrieved from huggingface; the
```
Comment (sarahnlewis) [inline src/bioemu/sample.py]: Is the biorxiv URL removed deliberately?

Is most of the diff in this docstring just new linebreaks?

File: src/bioemu/sample.py [hunk 18]
```diff
@@ -137,48 +217,99 @@ def main(
...
+        if steering_config_dict is not None:
+            description = f"Sampling batch {seed}/{num_samples} ({n} samples with {steering_config_dict['num_particles']} particles)"
+        else:
+            description = f"Sampling batch {seed}/{num_samples} ({n} samples)"
+        batch_iterator.set_description(description)
+
+        steering_config_dict = (
+            OmegaConf.create(steering_config_dict) if steering_config_dict is not None else None
+        )
+        print(
```
Comment (sarahnlewis) [inline src/bioemu/sample.py]: Remove this print statement or convert to logger.info and put it outside the loop over batches.

File: src/bioemu/sample.py [hunk 19]
```diff
@@ -188,9 +319,11 @@ def main(
     if set(sequences) != {sequence}:
         raise ValueError(f"Expected all sequences to be {sequence}, but got {set(sequences)}")
     positions = torch.tensor(np.concatenate([np.load(f)["pos"] for f in samples_files]))
+    # denoised_positions = torch.tensor(np.concatenate([np.load(f)["denoised_pos"] for f in samples_files], axis=0))
```
Comment (sarahnlewis) [inline src/bioemu/sample.py]: Remove commented code

File: src/bioemu/sample.py [hunk 20]
```diff
@@ -280,19 +421,24 @@ def generate_batch(
...
+        steering_config=steering_config,
     )
     assert isinstance(sampled_chemgraph_batch, Batch)
     sampled_chemgraphs = sampled_chemgraph_batch.to_data_list()
-    pos = torch.stack([x.pos for x in sampled_chemgraphs]).to("cpu")
-    node_orientations = torch.stack([x.node_orientations for x in sampled_chemgraphs]).to("cpu")
+    pos = torch.stack([x.pos for x in sampled_chemgraphs]).to("cpu")  # [BS, L, 3]
+    node_orientations = torch.stack([x.node_orientations for x in sampled_chemgraphs]).to(
+        "cpu"
+    )  # [BS, L, 3, 3]
```
Comment (sarahnlewis) [inline src/bioemu/sample.py]: thanks for adding the shape comments

File: src/bioemu/config/steering/physical_steering.yaml [hunk 21]
```diff
@@ -0,0 +1,48 @@
...
+
+num_particles: 5 # Number of particles for steering
+start: 0.1 # Start time for steering
+end: 0.0 # End time for steering (default: continue until end)
+resampling_freq: 5 # Resample every N steps
+potentials:
+  chainbreak:
+    # Enforces realistic Ca-Ca distances (3.8A) using flat-bottom loss
+    _target_: bioemu.steering.ChainBreakPotential
+    # flatbottom: Minimum allowed distance between non-neighboring Ca atoms (A) - zero penalty within this range
```
Comment (sarahnlewis) [inline src/bioemu/config/steering/physical_steering.yaml]:
```suggestion
    # flatbottom: Maximum allowed distance between neighboring Ca atoms (A) - zero penalty within this range
```
?

Actually, I'd suggest to leave these comments out of this file and just document the parameters in the class definitions for ChainBreakPotential etc. so you only have to maintain docs in one place.

File: src/bioemu/steering.py [hunk 22]
```diff
@@ -0,0 +1,333 @@
...
+    sde: SDE,
+    R: torch.Tensor,
+    t: torch.Tensor,
+    batch_idx: torch.LongTensor,
+    score: torch.Tensor,
+) -> torch.Tensor:
+    """
+    Compute R_0 given R_t and score.
+    """
+    alpha_t, sigma_t = sde.mean_coeff_and_std(x=R, t=t, batch_idx=batch_idx)
```
Comment (sarahnlewis) [inline src/bioemu/steering.py]: hm, is this legit when R lives in SO(3)? What do we need it for anyway? I thought the potentials were all defined in terms of CA positions.

File: notebooks/disulfide_bridge_steering_example.py
Comment (ludwigwinkler) [inline notebooks/disulfide_bridge_steering_example.py]: Quick numerical test to check disulfide bridge steering.

File: pyproject.toml [hunk 23]
```diff
@@ -22,6 +22,9 @@ dependencies = [
     "stackprinter",
     "typer",
     "uv",
+    "einops",
+    "matplotlib>=3.10.7",
```
[the remark this answers is not in the document]
Comment (ludwigwinkler) [reply pyproject.toml]: the ppft notebook is using matplotib and my own steering example notebook does as well. Pandas is not required and I removed it.

File: src/bioemu/config/steering/physical_steering.yaml (hunk 21 above)
Comment (ludwigwinkler) [reply src/bioemu/config/steering/physical_steering.yaml]: moved it to the class definition

File: src/bioemu/denoiser.py [hunk 24]
```diff
@@ -414,4 +432,67 @@ def dpm_solver(
...
+                atom37[..., 4, :],
+            )  # [BS, L, 4, 3] -> [BS, L, 3] for N,Ca,C,O
+            energies = []
+            for potential_ in fk_potentials:
+                energies += [potential_(N_pos, Ca_pos, C_pos, O_pos, i=i, N=N)]
+                # energies += [potential_(None, 10 * x0_t, None, None, i=i, N=N)]
+
+            total_energy = torch.stack(energies, dim=-1).sum(-1)  # [BS]
+
+            # if resampling implicitely given by num_fk_samples > 1
```
Comment (sarahnlewis) [inline src/bioemu/denoiser.py]:
```suggestion
```

File: src/bioemu/denoiser.py [hunk 25]
```diff
@@ -414,4 +432,67 @@ def dpm_solver(
...
+                        energy=total_energy,
+                        previous_energy=previous_energy,
+                        num_particles=steering_config["num_particles"],
+                        log_weights=log_weights,
+                    )
+                    previous_energy = total_energy
+
+                # ... or a single final sample
+                elif i >= N - 2:  # The last step is N-2
+                    print(
```
Comment (sarahnlewis) [inline src/bioemu/denoiser.py]: use logger

File: src/bioemu/denoiser.py [hunk 26]
```diff
@@ -1,16 +1,18 @@
...
 import numpy as np
 import torch
 from torch_geometric.data.batch import Batch
+from tqdm.auto import tqdm
 
-from .chemgraph import ChemGraph
-from .sde_lib import SDE, CosineVPSDE
-from .so3_sde import SO3SDE, apply_rotvec_to_rotmat
-
-TwoBatches = tuple[Batch, Batch]
```
Comment (sarahnlewis) [inline src/bioemu/denoiser.py]: ha good spot

**Comment by ludwigwinkler** [comment]
@microsoft-github-policy-service agree company="Microsoft"

**Review by nw13slx — APPROVED** [review]
LGTM
